src/netconf_acm.c

```diff
@@ -0,0 +1,1862 @@
+}
+
+API void
+sr_nacm_destroy(void)
+{
+    struct sr_nacm_group *group;
+    struct sr_nacm_rule_list *rule_list, *tmp;
+    uint32_t i, j;
+
+    for (i = 0; i < nacm.group_count; ++i) {
+        group = &nacm.groups[i];
+        free(group->name);
+        for (j = 0; j < group->user_count; ++j) {
+            free(group->users[j]);
+        }
+        free(group->users);
+    }
+    free(nacm.groups);
+
+    LY_LIST_FOR_SAFE(nacm.rule_lists, tmp, rule_list) {
+        free(rule_list->name);
+        for (i = 0; i < rule_list->group_count; ++i) {
+            free(rule_list->groups[i]);
+        }
+        free(rule_list->groups);
+        sr_nacm_remove_rules(rule_list);
+        free(rule_list);
+    }
+
+    nacm.rule_lists = NULL;
+    nacm.groups = NULL;
+    nacm.group_count = 0;
+    nacm.denied_notifications = 0;
+    nacm.denied_operations = 0;
+    nacm.denied_data_writes = 0;
+    pthread_mutex_destroy(&nacm.lock);
+}
+
+/**
+ * @brief Get passwd entry of a user, specifically its UID and GID.
+ *
+ * @param[in] user User to learn about.
+ * @param[out] uid User UID, if set.
+ * @param[out] gid User GID, if set.
+ * @param[out] found 1 if user found, 0 if user not found
+ * @return err_info, NULL on success.
+ */
+static sr_error_info_t *
+sr_nacm_getpwnam(const char *user, uid_t *uid, gid_t *gid, int* found)
+{
+    sr_error_info_t *err_info = NULL;
+    struct passwd pwd, *pwd_p;
+    char *buf = NULL;
+    ssize_t buflen;
+    int ret;
+
+    assert(user);
+
+    buflen = sysconf(_SC_GETPW_R_SIZE_MAX);
+    if (buflen == -1) {
+        buflen = 2048;
+    }
+    buf = malloc(buflen);
+    if (!buf) {
+        SR_ERRINFO_MEM(&err_info);
+        return err_info;
+    }
+    ret = getpwnam_r(user, &pwd, buf, buflen, &pwd_p);
+    if (ret) {
+        /* If we got an error, we don't need to set `found`. */
+        sr_errinfo_new(&err_info, SR_ERR_OPERATION_FAILED, "Getting user \"%s\" pwd entry failed (%s).", user, strerror(ret));
+        free(buf);
+        return err_info;
+    } else if (!pwd_p) {
+        *found = 0;
+        free(buf);
+        return err_info;
+    }
+
+    *found = 1;
+
+    if (uid) {
+        *uid = pwd.pw_uid;
+    }
+    if (gid) {
+        *gid = pwd.pw_gid;
+    }
+    free(buf);
+    return err_info;
+}
+
+/**
+ * @brief Check NACM acces for the data tree. If this check passes, no other check is necessary.
+ * If not, each node must be checked separately to decide.
+ *
+ * @param[in] root Root schema node of the data subtree.
+ * @param[in] user User, whose access to check.
+ * @param[out] allowed 1 if access allowed, 0 if more checks are required.
+ * @return errinfo, NULL on success.
+ */
+static sr_error_info_t *
+sr_nacm_allowed_tree(const struct lysc_node *root, const char *user, int* allowed)
+{
+    int user_found;
+    sr_error_info_t *err_info = NULL;
+    uid_t user_uid;
+
+    /* 1) NACM is off */
+    if (!nacm.enabled) {
+        *allowed = 1;
+        return err_info;
```

Comment:
If returning always on success I prefer just `return NULL;`, it makes the fact obvious.